### DISCUSSION TITLE
Smart-account deployment: details

### DISCUSSION THREAD

**Discussion by yogh333** [LedgerHQ/ledger-fresh-management#61] [Specifications]
> [Deploy Fresh Account](https://lucid.app/lucidchart/28c9816f-92a9-451f-a919-87c7db925932/edit?invitationId=inv_8d097081-4256-43d1-994b-810fbfa5aadb)
> 
> Regarding this sequence, I have already several questions:
> 
> 1. What kind of WebAuthn authenticators do we plan to support?
> 2. I have made the assumption that [Parsing and Validating the Registration Data](https://webauthn.guide/#about-webauthn) will be performed by the Deployer backend. Is it correct?
> 3. When the Deployer backend shall deploy an account contract, does it rely on [UDC](https://community.starknet.io/t/universal-deployer-contract-proposal/1864)?
> 4. How is the first plugin (WebAuthn) deployed? I have made the assumption that it is the Deployer but how the Deployer, who has its own account contract, can add a plugin to an other account? On the other side, when an account is first deplyed without any plugin (i.e without any public key), how can we interact with it ?
> 5. Account address shall be recoverable as when the end user will reconnect later using the same WebAuthn authenticator, an new account shall not be deployed. How do we plan to store the account address (deterministic address?) ?

Related to questions 3 and 4, it is yet not clear for me how the initial deployment is performed onchain. Below is a seq diagram describing on-chain operation, but I am not sure I have properly understood (DECLARE vs INVOKE tx, Deployer account vs Fresh account):

![Account Deployment (on chain)](https://user-images.githubusercontent.com/12996226/208084166-57fb24ce-1399-4625-acbd-f1d57fa41da7.png)

Would be happy to get your feedback !

_Originally posted by @yogh333 in https://github.com/LedgerHQ/ledger-fresh-management/issues/21#issuecomment-1354557714_

**Comment by CremaFR** [comment]
2. I guess Yes
3. Yes
4.

The account is deployed through a proxy, this imply you deploy the [proxy class hash](https://github.com/argentlabs/starknet-plugin-account/blob/main/contracts/upgrade/Proxy.cairo), not the fresh account clash.
The proxy has a `constructor`, this contrustor takes the fresh account clash **and** params to call a function on the implementation, here `initialize`. And initialize adds the first plugin

An example can be found [here, in the tests](https://github.com/argentlabs/starknet-plugin-account/blob/main/tests/test_account.cairo#L22)

ps: I would split the declare and deploy phases. Declare is done only once per contract and it is just a requirement for a deployment.

**Comment by yogh333** [comment]
@CremaFR Thanks for the clarification. I agree about _declare_ phase, put them in the diagram just as a prerequisite. I have updated the on-chain operation sequence diagram as following:
![Account Deployment (on chain) (1)](https://user-images.githubusercontent.com/12996226/208120733-997a098c-f62f-4c05-9966-e2ca91d7eb9f.png)

**Comment by yogh333** [comment]
Here are the updated sequence diagrams:
![Deploy Fresh Account](https://user-images.githubusercontent.com/12996226/208130521-1abda9e9-5503-4ad2-8f70-5d2a15204a1a.png)
![Account Deployment (on chain)](https://user-images.githubusercontent.com/12996226/208130544-a23b2322-29c7-4280-a6bd-a25e78794779.png)

**Comment by yogh333** [comment]
Authentication step:
![Authentication](https://user-images.githubusercontent.com/12996226/208397925-add1205f-0556-4061-a5ec-bd9a55c9cb9e.png)

**Comment by yogh333** [comment]
Linked with #7

**Comment by yogh333** [comment]
⚠️ (in french sorry): Il existe en fait depuis la 0.10.1 un [nouveau type de Tx](https://github.com/starkware-libs/cairo-lang/releases/tag/v0.10.1) `DeployAccount` qui permet de déployer un compte et de lui faire payer les frais. Suite à la disparition de la Tx `Deploy` il existe donc 2 manières de déployer un compte:

1. Un account a déjà été déployé et possède un fonction qui permet de déployer des contrats (via un appel au syscall `deploy`): dans ce cas, pour déployer un nouveau compte, il suffit d’un transaction `Invoke` sur l’account lui-même qui appelle cette fonction de déploiement de contrat. Les fees seront payés par l’account qui déploie (deployer).
2. On n’a pas d’account qui permet de déployer un contrat. Dans ce cas, on utilise une Tx `DeployAccount` qui permet d’auto-deployer un compte et de payer les fees associés. Il faut au préalable que l’adresse du compte (prédéterminée) soit approvisionnée en ETH.

L’utilisation de l’[Universal Deployer Contract](https://community.starknet.io/t/universal-deployer-contract-proposal/1864) rejoint le cas 1 mais permet de centraliser le code de déploiement de contrat dans un seul contrat plutôt que dans chaque account contract. Ansi, pour déployer un contrat, il suffit de faire une `Invoke` Tx vers l’UDC et sa fonction `deploy_contract` (plus besoin d’une fonction de déploiement de contrat dans un account contract).

---


Translated 🇬🇧

Since 0.10.1 there is a [new type of Tx](https://github.com/starkware-libs/cairo-lang/releases/tag/v0.10.1) `DeployAccount` that allows you to deploy an account and charge it. Following the disappearance of the Tx `Deploy` there are 2 ways to deploy an account:

1. An account has already been deployed and has a function that allows you to deploy contracts (via a call to syscall `deploy`): in this case, to deploy a new account, you only need an `Invoke` transaction on the account itself that calls this contract deployment function. The fee will be paid by the deploying account.

2. We do not have an account that deploys a contract. In this case, we use a Tx `DeployAccount` that automatically deploys an account and pays the associated fees. It is necessary to have a predefined address for the account in ETH.

Using the [Universal Deployer contract](https://community.starknet.io/t/universal-deployer-contract-proposal/1864) is similar to case 1 but allows the contract deployment code to be centralized in a single contract rather than in each account contract. Thus, to deploy a contract, you only need to do an `Invoke` Tx to the UDC and its `deploy_contract` function (no need for a contract deployment function in an account contract).
